internal/pkg/runtime/engines/singularity/container.go

```diff
@@ -994,14 +1007,28 @@ func (c *container) addDevMount(system *mount.System) error {
 			devptsPath, _ := c.session.GetPath("/dev/pts")
 			err = system.Points.AddFS(mount.DevTag, devptsPath, "devpts", syscall.MS_NOSUID|syscall.MS_NOEXEC, options)
 			if err != nil {
-				sylog.Verbosef("Couldn't mount devpts filesystem, continuing with PTY functionality disabled")
-			} else {
-				if err := c.addSessionDev("/dev/tty", system); err != nil {
+				return fmt.Errorf("failed to add devpts filesystem: %s", err)
+			}
+			// add additional PTY allocation symlink
+			if err := c.session.AddSymlink("/dev/ptmx", "/dev/pts/ptmx"); err != nil {
+				return fmt.Errorf("failed to create /dev/ptmx symlink: %s", err)
+			}
+
+		}
+		// add /dev/console mount pointing to original tty if there is one
+		for fd := 0; fd <= 2; fd++ {
```

Comment:
To continue with the first comment, you could bind there instead, no need to call ttyname, just call `https://godoc.org/golang.org/x/crypto/ssh/terminal#IsTerminal` and bind accordingly with `/proc/self/fd/X` to `/dev/console`

Reply:
What you mean is to do a readlink() on /proc/self/fd/X and bind that.  The problem is that doesn't always work, for example inside docker where that mysteriously points to non-existent "/N" paths.   ttyname() has a better chance of working when the /proc/self/fd/X symlink targets are broken.

Reply:
In slack conversation with @cclerget on this, the real thing he'd like to avoid is calling a C function from go.  Unfortunately there's no go version of ttyname (other than one we found that was covered by GPL) so for now he asked me to replace this code with the less functional readlink().  He will later as part of the OCI implementation change it to invoke ttyname() in starter.c.